plugins/catalog-backend/src/ingestion/processors/gitlab/groups.ts

```diff
@@ -0,0 +1,179 @@
+ * limitations under the License.
+ */
+import { GroupEntity } from '@backstage/catalog-model';
+import { GitLabClient, paginated } from './client';
+import { getGroupMembers } from './users';
+
+export type GroupNode = {
+  entity: GroupEntity;
+  parent: number | null;
+  children: number[];
+};
+
+export type GroupAdjacency = Map<number, GroupNode>;
+
+type Group = {
+  id: number;
+  web_url: string;
+  name: string;
+  path: string;
+  description: string;
+  full_name: string;
+  full_path: string;
+  created_at: string;
+  parent_id: number | null;
+};
+
+type SharedGroup = {
+  group_id: number;
+  group_name: string;
+  group_full_path: string;
+  group_access_level: number;
+  expires_at?: string;
+};
+
+export async function getGroups(
+  client: GitLabClient,
+  _id: string,
+  pathDelimiter: string,
+  groupType: string = 'team',
+): Promise<GroupAdjacency> {
+  const groups = paginated<Group>(
+    options => client.pagedRequest('/groups', options),
+    { per_page: 100 },
+  );
+
+  const groupAdjacency = new Map<number, GroupNode>();
+  for await (const result of groups) {
+    const entity: GroupEntity = {
+      apiVersion: 'backstage.io/v1alpha1',
+      kind: 'Group',
+      metadata: {
+        name: result.full_path.replaceAll('/', pathDelimiter),
+      },
+      spec: {
+        type: groupType,
+        profile: {},
+        children: [],
+        members: [],
+      },
+    };
+    if (result.name) entity.spec!.profile!.displayName = result.name;
+    if (result.description) entity.metadata!.description = result.description;
+
+    if (!groupAdjacency.has(result.id)) {
+      groupAdjacency.set(result.id, {
+        children: [],
+        parent: result.parent_id,
+        entity,
+      });
+    }
+  }
+  await populateChildrenMembers(client, groupAdjacency);
+  mapChildrenToEntityRefs(groupAdjacency);
+  return groupAdjacency;
+}
+
+export async function populateChildrenMembers(
+  client: GitLabClient,
+  groupAdjacency: GroupAdjacency,
+) {
+  // map ids of children using parent for each group node
+  for (const [id, { parent, entity }] of groupAdjacency) {
+    // append as child of parent if present
+    if (parent !== null) {
+      const parentGroupNode = groupAdjacency.get(parent);
+      if (parentGroupNode) {
+        parentGroupNode.children.push(id);
+      }
+    }
+    // populate direct members
+    const users = await getGroupMembers(client, String(id), false);
+    for (const user of users) {
+      entity.spec!.members!.push(`user:${user.metadata.name}`);
+    }
+    // populate direct members of shared_with_groups
+    const sharedWithGroups = await getSharedWithGroupsIDs(client, String(id));
+    for (const sharedID of sharedWithGroups) {
+      // populate direct members
+      const sharedGroupUsers = await getGroupMembers(client, sharedID, false);
+      for (const user of sharedGroupUsers) {
+        entity.spec!.members!.push(`user:${user.metadata.name}`);
```

Comment:
No problems with duplicate members here? Or in general, in other parts of the code.

Reply:
Yes, duplicate members can be an issue. I'm going to port the processor to the provider pattern and lift its config up (similar to the ldap provider module) so we can resolve duplicates before emitting.